# Assignment 1
## Part 1 - Theory Questions

### k-Nearest Neighbor Classification
Q1: Assume that you have a large training dataset. Specify a disadvantage of the k-Nearest Neighbor method when using it during testing. State also your reason about your answer. 
- Ans: During Testing K-Nearest Neighbor method requires O(N) computational power, and that means when using a large scale dataset such as million or billion scaled, we need to compare the query with all the data in there and it is not efficient.
  
Q2: Considering the image below, state an optimal k-value depending on that the algorithm you are using is k-Nearest Neighbor. State also your reason behind the optimal value you preferred.

![imgx](pictures/p1_1.png)

- Ans: The preffered optimal value is 12 according to the image, as validation error is minimum there, we can conclude that optimal k value is 12

![Figure 1](pictures/p1_2.png)

Q3: Assume that you have the following training set of positive (+), negative (-) instances and a single test instance (o) in the image below (Figure 1). Assume also that the Euclidean metric is used for measuring the distance between instances. Finally consider that every nearest neighbor instance affects the final vote equally.

- What is the class appointed to the test instance for K=1? State also reason behind your answer.
    - In figure 1 we can say that test instance's most clossest K point is -, so the test instance should be - as every nearest neighbor instance affects the final vote equally.
- What is the class appointed to the test instance for K=3? State also reason behind your answer.
    - In figure 1 we can say that test instance's most clossest K point is (-, -, +), so the test instance should be - as every nearest neighbor instance affects the final vote equally.
- What is the class appointed to the test instance for K=5? State also reason behind your answer.
    - In figure 1 we can say that test instance's most clossest K point is (-, -, +, +, +), so the test instance should be + as every nearest neighbor instance affects the final vote equally.

Q4: Fill the blanks with T (True) or F (False) for the statements below:

- instances of the data have the same scale then k-Nearest Neighbor’s performance increases drastically. (T)
- While k-Nearest Neighbor performs well with a small number of input variables, it’s performance decreases when the number of inputs becomes large. (T)
-  k-Nearest Neighbor makes supposes nothing about the functional form of the problem it handles. (T)




### Linear Regression

Q1:  Assume that you have five students have registered to a class and the class have a midterm and the final exam. You have obtained a set of their marks on two exams, which is in the table below:

![linear1](pictures/p1_3.png)
You plan to a model which form’s is fθ(x) = θ0 +θ1x1 +θ2x2 for fitting the data above. The x1 shows midterm exam score while x2 shows square of the midterm score. Besides you plan to use feature scaling (using divide operation by the ”max-min”, or range, of a feature) and mean normalization. What is the normalized value of the feature x(5)2?

- Ans: As x2 shows Square of Midterm score, x(5)2 should be 2025, but as there is a normalization part, and 2025 is the minimum value in Square of Midterm Scores, X(5)2 is now 0

Q2: Considering the figure below, which of the offsets used in linear regressions least square line fit? Assume that horizontal axis represents independent variable and vertical axis represents dependent variable. State your answer with your proper explanation.

![linear2](pictures/p1_4.png)

- Ans: As we are searching for best weights and biases for the data we have in Linear regression, we try to minimize the loss of predicted values for the data we have and orignal values. And that means we are looking into vertical offset as we try to minimize it.

Q3: Considering the table below, consisting of four training examples:

![linear3](pictures/p1_5.png)

Assume that you are trying to fit the data above to the linear regression model fθ(x) = θ0+θ1x1. Find the θ0 and θ1 values by using closed form solution (θ = (X^TX)**(-1) X^Ty) Also state dimension values of X, y, and θ matrices. Finally, show your calculations step by step.

- Ans: as X consists x0 (which is 1) and x1 values we can write X as;
X = [[1, 1], [1, 2], [1, 4], [1, 0]]

and that means;
X^T = [[1, 1, 1, 1], [1, 2, 4, 0]]

and that makes
X^TX = [[1, 1, 1, 1], [1, 2, 4, 0]] @ [[1, 1], [1, 2], [1, 4], [1, 0]] = [[4, 7], [7, 21]]

As the formula below says;

![linear4](pictures/p1_6.png)

inv(X^TX) = 1 / (4*21-7*7) * [[21, -7], [-7, 4]] => inv(X^TX) = 1 / (84-49) * [[21, -7], [-7, 4]] => 1 / 35 * [[21, -7], [-7, 4]] = [[3/5, -1/5], [-1/5, 4/35]]

inv(X^TX) = [[3/5, -1/5], [-1/5, 1/35]]
now calculate the X^Ty;

X^Ty = [[1, 1, 1, 1], [1, 2, 4, 0]] @ [0.5, 1, 2, 0] = [[3.5], [10.5]]

and when we multiplicate this matrices

inv(X^TX)X^Ty = θ = [[3/5, -1/5], [-1/5, 4/35]] @ [[3.5], [10.5]] = [3/5*3.5-1/5*10.5, -1/5*3.5+4/35*10.5] = [0, 0.5]

θ = [0, 0.5]

Q4: State a valid reason for feature scaling and explain why it is a valid reason with respect to your reasoning.

- Ans: There are several reasons why feature scaling is important;
1) Every feature can be different so one of them can be too big and that makes other features contribute less even if they are important.

2) It helps converge faster and finding global minimum faster.

# PART 2 - Anime Recommendation and Rating Prediction
#### In this Part We need to make a anime recommendation and rating prediction system using anime data and user data

### Installing Gdown to download files from gdrive

In [1]:
!pip install gdown

  Using cached gdown-4.7.1-py3-none-any.whl (15 kB)
  Using cached filelock-3.13.0-py3-none-any.whl.metadata (2.8 kB)
  Using cached requests-2.31.0-py3-none-any.whl.metadata (4.6 kB)
  Using cached tqdm-4.66.1-py3-none-any.whl.metadata (57 kB)
  Using cached beautifulsoup4-4.12.2-py3-none-any.whl (142 kB)
  Using cached soupsieve-2.5-py3-none-any.whl.metadata (4.7 kB)
  Using cached idna-3.4-py3-none-any.whl (61 kB)
  Using cached urllib3-2.0.7-py3-none-any.whl.metadata (6.6 kB)
  Using cached certifi-2023.7.22-py3-none-any.whl.metadata (2.2 kB)
  Using cached PySocks-1.7.1-py3-none-any.whl (16 kB)
Using cached filelock-3.13.0-py3-none-any.whl (11 kB)
Using cached tqdm-4.66.1-py3-none-any.whl (78 kB)
Using cached certifi-2023.7.22-py3-none-any.whl (158 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 137.5/137.5 kB 87.9 kB/s eta 0:00:00a 0:00:01
Using cached soupsieve-2.5-py3-none-any.whl (36 kB)
Using cached urllib3-2.0.7-py3-none-any.whl (124 kB)
Using cached requests-2.31.0-py3-none

In [2]:
!gdown --id 1kCL8dZLHQUlBUzH-DmYHRNi7xMLrgS4b

/home/grkmkola/miniconda3/envs/ml/lib/python3.11/site-packages/gdown/cli.py:126: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  warnings.warn(
Downloading...
From: https://drive.google.com/uc?id=1kCL8dZLHQUlBUzH-DmYHRNi7xMLrgS4b
To: /home/grkmkola/Desktop/Assignments/ML/Dataset_A1.zip
100%|██████████████████████████████████████| 1.25M/1.25M [00:14<00:00, 86.1kB/s]


In [4]:
# unzipping the folder , the contents are placed in Data->output->/kaggle/working 
# Currently each user is limited to 20GB data in kaggle 
! unzip Dataset_A1.zip
! rm Dataset_A1.zip

rm: cannot remove 'Dataset_A1.zip': No such file or directory


In [5]:
!pip install joblib numba

  Using cached joblib-1.3.2-py3-none-any.whl.metadata (5.4 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.2/61.2 kB 83.4 kB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.2/302.2 kB 76.3 kB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 84.3 kB/s eta 0:00:0000:0100:02
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.6/43.6 MB 81.1 kB/s eta 0:00:0000:0100:14m
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.2/18.2 MB 83.8 kB/s eta 0:00:0000:0100:06


#### Importing some important libraries

In [1]:
import pandas as pd
from PIL import Image
import numpy as np

#### Reading files and creating Pandas Dataframes for animes and users train and test data

In [2]:
anime_df = pd.read_csv('animes.csv').dropna()
train_df = pd.read_csv('user_rates_train.csv').dropna()
test_df = pd.read_csv('user_rates_test.csv').dropna()

### Preprocessing Anime Dataframe

In [3]:
# Replacing UNKNOWN to UnknownGenre
anime_df['Genres'] = anime_df['Genres'].replace('UNKNOWN', 'UnkownGenre').apply(lambda x: x.split(','))
anime_df

anime_id                             Name  \
0             1                     Cowboy Bebop   
1             5  Cowboy Bebop: Tengoku no Tobira   
2             6                           Trigun   
3             7               Witch Hunter Robin   
4             8                   Bouken Ou Beet   
...         ...                              ...   
24900     55731                      Wu Nao Monu   
24901     55732              Bu Xing Si: Yuan Qi   
24902     55733                      Di Yi Xulie   
24903     55734         Bokura no Saishuu Sensou   
24904     55735                   Shijuuku Nichi   

                                          Genres   Type         Studios  \
0              [Action,  Award Winning,  Sci-Fi]     TV         Sunrise   
1                              [Action,  Sci-Fi]  Movie           Bones   
2                  [Action,  Adventure,  Sci-Fi]     TV        Madhouse   
3      [Action,  Drama,  Mystery,  Supernatural]     TV         Sunrise   
4           [Adventure,  Fantasy,  Supernatural]     TV  Toei Animation   
...                                          ...    ...             ...   
24900         [Comedy,  Fantasy,  Slice of Life]    ONA         UNKNOWN   
24901             [Action,  Adventure,  Fantasy]    ONA         UNKNOWN   
24902    [Action,  Adventure,  Fantasy,  Sci-Fi]    ONA         UNKNOWN   
24903                              [UnkownGenre]  Music         UNKNOWN   
24904                              [UnkownGenre]  Music         UNKNOWN   

          Source       Duration  \
0       Original  24 min per ep   
1       Original    1 hr 55 min   
2          Manga  24 min per ep   
3       Original  25 min per ep   
4          Manga  23 min per ep   
...          ...            ...   
24900  Web manga        Unknown   
24901  Web novel        Unknown   
24902  Web novel        Unknown   
24903   Original          3 min   
24904   Original          3 min   

                                               Image URL  
0      https://cdn.myanimelist.net/images/anime/4/196...  
1      https://cdn.myanimelist.net/images/anime/1439/...  
2      https://cdn.myanimelist.net/images/anime/7/203...  
3      https://cdn.myanimelist.net/images/anime/10/19...  
4      https://cdn.myanimelist.net/images/anime/7/215...  
...                                                  ...  
24900  https://cdn.myanimelist.net/images/anime/1386/...  
24901  https://cdn.myanimelist.net/images/anime/1383/...  
24902  https://cdn.myanimelist.net/images/anime/1130/...  
24903  https://cdn.myanimelist.net/images/anime/1931/...  
24904  https://cdn.myanimelist.net/images/anime/1902/...  

[24905 rows x 8 columns]

In [4]:
genres = set()
anime_df['Genres'].apply(lambda x: genres.update(x))

for genre in genres:
    anime_df[genre] = 0
    

for i, genre in enumerate(anime_df['Genres']):
    for g in genre:
        anime_df.loc[i, g] = 1 
    
anime_df.drop('Genres', axis=1, inplace=True)
anime_df

anime_id                             Name   Type         Studios  \
0             1                     Cowboy Bebop     TV         Sunrise   
1             5  Cowboy Bebop: Tengoku no Tobira  Movie           Bones   
2             6                           Trigun     TV        Madhouse   
3             7               Witch Hunter Robin     TV         Sunrise   
4             8                   Bouken Ou Beet     TV  Toei Animation   
...         ...                              ...    ...             ...   
24900     55731                      Wu Nao Monu    ONA         UNKNOWN   
24901     55732              Bu Xing Si: Yuan Qi    ONA         UNKNOWN   
24902     55733                      Di Yi Xulie    ONA         UNKNOWN   
24903     55734         Bokura no Saishuu Sensou  Music         UNKNOWN   
24904     55735                   Shijuuku Nichi  Music         UNKNOWN   

          Source       Duration  \
0       Original  24 min per ep   
1       Original    1 hr 55 min   
2          Manga  24 min per ep   
3       Original  25 min per ep   
4          Manga  23 min per ep   
...          ...            ...   
24900  Web manga        Unknown   
24901  Web novel        Unknown   
24902  Web novel        Unknown   
24903   Original          3 min   
24904   Original          3 min   

                                               Image URL  Girls Love  Gourmet  \
0      https://cdn.myanimelist.net/images/anime/4/196...           0        0   
1      https://cdn.myanimelist.net/images/anime/1439/...           0        0   
2      https://cdn.myanimelist.net/images/anime/7/203...           0        0   
3      https://cdn.myanimelist.net/images/anime/10/19...           0        0   
4      https://cdn.myanimelist.net/images/anime/7/215...           0        0   
...                                                  ...         ...      ...   
24900  https://cdn.myanimelist.net/images/anime/1386/...           0        0   
24901  https://cdn.myanimelist.net/images/anime/1383/...           0        0   
24902  https://cdn.myanimelist.net/images/anime/1130/...           0        0   
24903  https://cdn.myanimelist.net/images/anime/1931/...           0        0   
24904  https://cdn.myanimelist.net/images/anime/1902/...           0        0   

        Award Winning  ...  Sports   Boys Love   Comedy   Drama  Boys Love  \
0                   1  ...       0           0        0       0          0   
1                   0  ...       0           0        0       0          0   
2                   0  ...       0           0        0       0          0   
3                   0  ...       0           0        0       1          0   
4                   0  ...       0           0        0       0          0   
...               ...  ...     ...         ...      ...     ...        ...   
24900               0  ...       0           0        0       0          0   
24901               0  ...       0           0        0       0          0   
24902               0  ...       0           0        0       0          0   
24903               0  ...       0           0        0       0          0   
24904               0  ...       0           0        0       0          0   

        Supernatural  Supernatural   Girls Love  Mystery  UnkownGenre  
0                  0             0            0        0            0  
1                  0             0            0        0            0  
2                  0             0            0        0            0  
3                  1             0            0        0            0  
4                  1             0            0        0            0  
...              ...           ...          ...      ...          ...  
24900              0             0            0        0            0  
24901              0             0            0        0            0  
24902              0             0            0        0            0  
24903              0             0            0        

In [5]:
# Replace 'UNKNOWN' with 'UnknownType'
anime_df['Type'] = anime_df['Type'].replace('UNKNOWN', 'UnkownType')

# Get one-hot encoding of 'Type'
one_hot = pd.get_dummies(anime_df['Type']).astype(int)

# Drop column 'Type' as it is now encoded
anime_df.drop('Type', axis=1, inplace=True)
anime_df = anime_df.join(one_hot)

anime_df

anime_id                             Name         Studios     Source  \
0             1                     Cowboy Bebop         Sunrise   Original   
1             5  Cowboy Bebop: Tengoku no Tobira           Bones   Original   
2             6                           Trigun        Madhouse      Manga   
3             7               Witch Hunter Robin         Sunrise   Original   
4             8                   Bouken Ou Beet  Toei Animation      Manga   
...         ...                              ...             ...        ...   
24900     55731                      Wu Nao Monu         UNKNOWN  Web manga   
24901     55732              Bu Xing Si: Yuan Qi         UNKNOWN  Web novel   
24902     55733                      Di Yi Xulie         UNKNOWN  Web novel   
24903     55734         Bokura no Saishuu Sensou         UNKNOWN   Original   
24904     55735                   Shijuuku Nichi         UNKNOWN   Original   

            Duration                                          Image URL  \
0      24 min per ep  https://cdn.myanimelist.net/images/anime/4/196...   
1        1 hr 55 min  https://cdn.myanimelist.net/images/anime/1439/...   
2      24 min per ep  https://cdn.myanimelist.net/images/anime/7/203...   
3      25 min per ep  https://cdn.myanimelist.net/images/anime/10/19...   
4      23 min per ep  https://cdn.myanimelist.net/images/anime/7/215...   
...              ...                                                ...   
24900        Unknown  https://cdn.myanimelist.net/images/anime/1386/...   
24901        Unknown  https://cdn.myanimelist.net/images/anime/1383/...   
24902        Unknown  https://cdn.myanimelist.net/images/anime/1130/...   
24903          3 min  https://cdn.myanimelist.net/images/anime/1931/...   
24904          3 min  https://cdn.myanimelist.net/images/anime/1902/...   

       Girls Love  Gourmet   Award Winning   Fantasy  ...   Girls Love  \
0               0        0               1         0  ...            0   
1               0        0               0         0  ...            0   
2               0        0               0         0  ...            0   
3               0        0               0         0  ...            0   
4               0        0               0         1  ...            0   
...           ...      ...             ...       ...  ...          ...   
24900           0        0               0         1  ...            0   
24901           0        0               0         1  ...            0   
24902           0        0               0         1  ...            0   
24903           0        0               0         0  ...            0   
24904           0        0               0         0  ...            0   

       Mystery  UnkownGenre  Movie  Music  ONA  OVA  Special  TV  UnkownType  
0            0            0      0      0    0    0        0   1           0  
1            0            0      1      0    0    0        0   0           0  
2            0            0      0      0    0    0        0   1           0  
3            0            0      0      0    0    0        0   1           0  
4            0            0      0      0    0    0        0   1           0  
...        ...          ...    ...    ...  ...  ...      ...  ..         ...  
24900        0            0      0      0    1    0        0   0           0  
24901        0            0      0      0    1    0        0   0           0  
24902        0            0      0      0    1    0        0   0           0  
24903        0            1      0      1    0    0        0   0           0  
24904        0            1      0      1    0    0        0   0           0  

[24905 rows x 55 columns]

In [6]:

# # As we can drop Studios column with the new update, I decided to drop it as it makes complexity much higher
# # Replace 'UNKNOWN' with 'UnknownStudio'
# anime_df['Studios'] = anime_df['Studios'].replace('UNKNOWN', 'UnkownStudios')

# # Get one-hot encoding of 'Studios'
# one_hot = pd.get_dummies(anime_df['Studios']).astype(int)

# # Drop column 'Studios' as it is now encoded
# anime_df.drop('Studios', axis=1, inplace=True)
# anime_df = anime_df.join(one_hot)

# anime_df

anime_df.drop('Studios', axis=1, inplace=True)
anime_df

anime_id                             Name     Source       Duration  \
0             1                     Cowboy Bebop   Original  24 min per ep   
1             5  Cowboy Bebop: Tengoku no Tobira   Original    1 hr 55 min   
2             6                           Trigun      Manga  24 min per ep   
3             7               Witch Hunter Robin   Original  25 min per ep   
4             8                   Bouken Ou Beet      Manga  23 min per ep   
...         ...                              ...        ...            ...   
24900     55731                      Wu Nao Monu  Web manga        Unknown   
24901     55732              Bu Xing Si: Yuan Qi  Web novel        Unknown   
24902     55733                      Di Yi Xulie  Web novel        Unknown   
24903     55734         Bokura no Saishuu Sensou   Original          3 min   
24904     55735                   Shijuuku Nichi   Original          3 min   

                                               Image URL  Girls Love  Gourmet  \
0      https://cdn.myanimelist.net/images/anime/4/196...           0        0   
1      https://cdn.myanimelist.net/images/anime/1439/...           0        0   
2      https://cdn.myanimelist.net/images/anime/7/203...           0        0   
3      https://cdn.myanimelist.net/images/anime/10/19...           0        0   
4      https://cdn.myanimelist.net/images/anime/7/215...           0        0   
...                                                  ...         ...      ...   
24900  https://cdn.myanimelist.net/images/anime/1386/...           0        0   
24901  https://cdn.myanimelist.net/images/anime/1383/...           0        0   
24902  https://cdn.myanimelist.net/images/anime/1130/...           0        0   
24903  https://cdn.myanimelist.net/images/anime/1931/...           0        0   
24904  https://cdn.myanimelist.net/images/anime/1902/...           0        0   

        Award Winning   Fantasy   Sports  ...   Girls Love  Mystery  \
0                   1         0        0  ...            0        0   
1                   0         0        0  ...            0        0   
2                   0         0        0  ...            0        0   
3                   0         0        0  ...            0        0   
4                   0         1        0  ...            0        0   
...               ...       ...      ...  ...          ...      ...   
24900               0         1        0  ...            0        0   
24901               0         1        0  ...            0        0   
24902               0         1        0  ...            0        0   
24903               0         0        0  ...            0        0   
24904               0         0        0  ...            0        0   

       UnkownGenre  Movie  Music  ONA  OVA  Special  TV  UnkownType  
0                0      0      0    0    0        0   1           0  
1                0      1      0    0    0        0   0           0  
2                0      0      0    0    0        0   1           0  
3                0      0      0    0    0        0   1           0  
4                0      0      0    0    0        0   1           0  
...            ...    ...    ...  ...  ...      ...  ..         ...  
24900            0      0      0    1    0        0   0           0  
24901            0      0      0    1    0        0   0           0  
24902            0      0      0    1    0        0   0           0  
24903            1      0      1    0    0        0   0           0  
24904            1      0      1    0    0        0   0           0  

[24905 rows x 54 columns]

In [7]:
# Replace 'UNKNOWN' with 'UnknownSource'
anime_df['Source'] = anime_df['Source'].replace('Music', 'MusicSource')

# Get one-hot encoding of 'Source'
one_hot = pd.get_dummies(anime_df['Source']).astype(int)

# Drop column 'Source' as it is now encoded
anime_df.drop('Source', axis=1, inplace=True)
anime_df = anime_df.join(one_hot)

anime_df

anime_id                             Name       Duration  \
0             1                     Cowboy Bebop  24 min per ep   
1             5  Cowboy Bebop: Tengoku no Tobira    1 hr 55 min   
2             6                           Trigun  24 min per ep   
3             7               Witch Hunter Robin  25 min per ep   
4             8                   Bouken Ou Beet  23 min per ep   
...         ...                              ...            ...   
24900     55731                      Wu Nao Monu        Unknown   
24901     55732              Bu Xing Si: Yuan Qi        Unknown   
24902     55733                      Di Yi Xulie        Unknown   
24903     55734         Bokura no Saishuu Sensou          3 min   
24904     55735                   Shijuuku Nichi          3 min   

                                               Image URL  Girls Love  Gourmet  \
0      https://cdn.myanimelist.net/images/anime/4/196...           0        0   
1      https://cdn.myanimelist.net/images/anime/1439/...           0        0   
2      https://cdn.myanimelist.net/images/anime/7/203...           0        0   
3      https://cdn.myanimelist.net/images/anime/10/19...           0        0   
4      https://cdn.myanimelist.net/images/anime/7/215...           0        0   
...                                                  ...         ...      ...   
24900  https://cdn.myanimelist.net/images/anime/1386/...           0        0   
24901  https://cdn.myanimelist.net/images/anime/1383/...           0        0   
24902  https://cdn.myanimelist.net/images/anime/1130/...           0        0   
24903  https://cdn.myanimelist.net/images/anime/1931/...           0        0   
24904  https://cdn.myanimelist.net/images/anime/1902/...           0        0   

        Award Winning   Fantasy   Sports   Adventure  ...  MusicSource  Novel  \
0                   1         0        0           0  ...            0      0   
1                   0         0        0           0  ...            0      0   
2                   0         0        0           1  ...            0      0   
3                   0         0        0           0  ...            0      0   
4                   0         1        0           0  ...            0      0   
...               ...       ...      ...         ...  ...          ...    ...   
24900               0         1        0           0  ...            0      0   
24901               0         1        0           1  ...            0      0   
24902               0         1        0           1  ...            0      0   
24903               0         0        0           0  ...            0      0   
24904               0         0        0           0  ...            0      0   

       Original  Other  Picture book  Radio  Unknown  Visual novel  Web manga  \
0             1      0             0      0        0             0          0   
1             1      0             0      0        0             0          0   
2             0      0             0      0        0             0          0   
3             1      0             0      0        0             0          0   
4             0      0             0      0        0             0          0   
...         ...    ...           ...    ...      ...           ...        ...   
24900         0      0             0      0        0             0          1   
24901         0      0             0      0        0             0          0   
24902         0      0             0      0        0             0          0   
24903         1      0             0      0        0             0          0   
24904         1      0             0      0        0             0          0   

       Web novel  
0              0  
1              0  
2              0  
3              0  
4              0  
...          ...  
24900          0  
24901          1  
24902          1  
24903          0  
24904          0  

[24905 rows x 70 columns]

In [8]:
def handle_duration(x):
    '''
        This Function handles duration and turn it into integer format in mins.
    '''
    mins = 0
    if "hr" in x:
        x = x.split("hr")
        mins = int(x[0])*60
        x = x[1]
    
    return int(x.split()[0]) + mins if "min" in x else mins

In [9]:
anime_df['Duration'] = anime_df['Duration'].apply(handle_duration)
maximum = np.max(anime_df['Duration'])
mean = np.mean(anime_df['Duration'])
anime_df['Duration'].fillna(mean, inplace=True)
anime_df['Duration'] /= maximum
anime_df

anime_id                             Name  Duration  \
0             1                     Cowboy Bebop  0.016000   
1             5  Cowboy Bebop: Tengoku no Tobira  0.076667   
2             6                           Trigun  0.016000   
3             7               Witch Hunter Robin  0.016667   
4             8                   Bouken Ou Beet  0.015333   
...         ...                              ...       ...   
24900     55731                      Wu Nao Monu  0.000000   
24901     55732              Bu Xing Si: Yuan Qi  0.000000   
24902     55733                      Di Yi Xulie  0.000000   
24903     55734         Bokura no Saishuu Sensou  0.002000   
24904     55735                   Shijuuku Nichi  0.002000   

                                               Image URL  Girls Love  Gourmet  \
0      https://cdn.myanimelist.net/images/anime/4/196...           0        0   
1      https://cdn.myanimelist.net/images/anime/1439/...           0        0   
2      https://cdn.myanimelist.net/images/anime/7/203...           0        0   
3      https://cdn.myanimelist.net/images/anime/10/19...           0        0   
4      https://cdn.myanimelist.net/images/anime/7/215...           0        0   
...                                                  ...         ...      ...   
24900  https://cdn.myanimelist.net/images/anime/1386/...           0        0   
24901  https://cdn.myanimelist.net/images/anime/1383/...           0        0   
24902  https://cdn.myanimelist.net/images/anime/1130/...           0        0   
24903  https://cdn.myanimelist.net/images/anime/1931/...           0        0   
24904  https://cdn.myanimelist.net/images/anime/1902/...           0        0   

        Award Winning   Fantasy   Sports   Adventure  ...  MusicSource  Novel  \
0                   1         0        0           0  ...            0      0   
1                   0         0        0           0  ...            0      0   
2                   0         0        0           1  ...            0      0   
3                   0         0        0           0  ...            0      0   
4                   0         1        0           0  ...            0      0   
...               ...       ...      ...         ...  ...          ...    ...   
24900               0         1        0           0  ...            0      0   
24901               0         1        0           1  ...            0      0   
24902               0         1        0           1  ...            0      0   
24903               0         0        0           0  ...            0      0   
24904               0         0        0           0  ...            0      0   

       Original  Other  Picture book  Radio  Unknown  Visual novel  Web manga  \
0             1      0             0      0        0             0          0   
1             1      0             0      0        0             0          0   
2             0      0             0      0        0             0          0   
3             1      0             0      0        0             0          0   
4             0      0             0      0        0             0          0   
...         ...    ...           ...    ...      ...           ...        ...   
24900         0      0             0      0        0             0          1   
24901         0      0             0      0        0             0          0   
24902         0      0             0      0        0             0          0   
24903         1      0             0      0        0             0          0   
24904         1      0             0      0        0             0          0   

       Web novel  
0              0  
1              0  
2              0  
3              0  
4              0  
...          ...  
24900          0  
24901          1  
24902          1  
24903          0  
24904          0  

[24905 rows x 70 columns]

In [10]:
anime_names = dict(zip(anime_df['anime_id'], anime_df['Name']))
train_user_names = dict(zip(train_df['user_id'], train_df['Username']))
test_user_names = dict(zip(test_df['user_id'], test_df['Username']))

recommend_X, recommend_y = anime_df.drop(['anime_id', 'Name', 'Image URL'], axis=1), anime_df['Name']

anime_df.drop(['Name', 'Image URL'], axis=1, inplace=True)
train_df.drop(['Username', 'Anime Title'], axis=1, inplace=True)
test_df.drop(['Username', 'Anime Title'], axis=1, inplace=True)

In [11]:
train_merged = train_df.merge(anime_df, on='anime_id')
test_merged = test_df.merge(anime_df, on='anime_id')
merged_df = pd.concat([train_merged, test_merged])

In [12]:
merged_df

user_id  anime_id  rating  Duration  Girls Love  Gourmet   Award Winning  \
0        549       790       9  0.016667           0        0               0   
1       3242       790       9  0.016667           0        0               0   
2       9377       790      10  0.016667           0        0               0   
3      11092       790       8  0.016667           0        0               0   
4      13980       790      10  0.016667           0        0               0   
..       ...       ...     ...       ...         ...      ...             ...   
872  1204987      3086       7  0.017333           0        0               0   
873  1204987       513       9  0.082667           0        0               1   
874  1257993     14741       8  0.016000           0        0               0   
875  1257993     13601      10  0.015333           0        0               0   
876  1257993      7311      10  0.107333           0        0               0   

      Fantasy   Sports   Adventure  ...  MusicSource  Novel  Original  Other  \
0           0        0           0  ...            0      0         1      0   
1           0        0           0  ...            0      0         1      0   
2           0        0           0  ...            0      0         1      0   
3           0        0           0  ...            0      0         1      0   
4           0        0           0  ...            0      0         1      0   
..        ...      ...         ...  ...          ...    ...       ...    ...   
872         1        0           0  ...            0      0         0      0   
873         1        0           0  ...            0      0         1      0   
874         0        0           0  ...            0      0         0      0   
875         0        0           0  ...            0      0         1      0   
876         0        0           0  ...            0      0         0      0   

     Picture book  Radio  Unknown  Visual novel  Web manga  Web novel  
0               0      0        0             0          0          0  
1               0      0        0             0          0          0  
2               0      0        0             0          0          0  
3               0      0        0             0          0          0  
4               0      0        0             0          0          0  
..            ...    ...      ...           ...        ...        ...  
872             0      0        0             0          0          0  
873             0      0        0             0          0          0  
874             0      0        0             0          0          0  
875             0      0        0             0          0          0  
876             0      0        0             0          0          0  

[50237 rows x 70 columns]

In [13]:
user_item_matrix = merged_df.pivot_table(index='user_id', columns='anime_id', values='rating')
user_item_matrix.fillna(0, inplace=True)
user_item_matrix /= 10
user_item_matrix

anime_id  1      5      6      7      8      15     16     17     18     \
user_id                                                                   
549         1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
644         1.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1519        0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2211        0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2297        0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
...         ...    ...    ...    ...    ...    ...    ...    ...    ...   
1289083     0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1289493     0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1289795     0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1289833     0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1290429     0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

anime_id  19     ...  50854  51145  51535  52034  52211  52273  52305  52865  \
user_id          ...                                                           
549         0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
644         0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1519        0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2211        0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2297        0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
...         ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
1289083     0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1289493     0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1289795     0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1289833     0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1290429     0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

anime_id  53148  53149  
user_id                 
549         0.0    0.0  
644         0.0    0.0  
1519        0.0    0.0  
2211        0.0    0.0  
2297        0.0    0.0  
...         ...    ...  
1289083     0.0    0.0  
1289493     0.0    0.0  
1289795     0.0    0.0  
1289833     0.0    0.0  
1290429     0.0    0.0  

[3050 rows x 3307 columns]

In [14]:
anime_df.set_index('anime_id', inplace=True)

In [15]:
# Using Item Based Matrix we can recommend Animes without training.
# We show users as Anime Features and that makes we can understand what genre,type,studio,source users like.
item_based_matrix = pd.DataFrame(columns=anime_df.columns)
for user in user_item_matrix.index:
    row = user_item_matrix.loc[user]
    non_zeros = row[row != 0]
    idxs = non_zeros.index.tolist()
    mean = anime_df.loc[idxs].mul(non_zeros.loc[idxs], axis=0).mean()
    item_based_matrix.loc[user] = mean
item_based_matrix

Duration  Girls Love  Gourmet   Award Winning   Fantasy   Sports  \
549      0.020783         0.0      0.0        0.145000  0.215000      0.0   
644      0.019618         0.0      0.0        0.193333  0.186667      0.0   
1519     0.023683         0.0      0.0        0.193750  0.106250      0.0   
2211     0.012917         0.0      0.0        0.100000  0.408333      0.0   
2297     0.017476         0.0      0.0        0.066667  0.313333      0.0   
...           ...         ...      ...             ...       ...      ...   
1289083  0.025971         0.0      0.0        0.116000  0.116000      0.0   
1289493  0.017714         0.0      0.0        0.000000  0.550000      0.0   
1289795  0.010616         0.0      0.0        0.000000  0.158824      0.0   
1289833  0.014627         0.0      0.0        0.000000  0.305000      0.0   
1290429  0.014558         0.0      0.0        0.000000  0.154545      0.0   

          Adventure    Horror  Award Winning   Erotica  ...  MusicSource  \
549        0.080000  0.000000       0.040000      0.00  ...          0.0   
644        0.193333  0.000000       0.000000      0.00  ...          0.0   
1519       0.093750  0.000000       0.000000      0.00  ...          0.0   
2211       0.275000  0.058333       0.029167      0.00  ...          0.0   
2297       0.033333  0.000000       0.000000      0.00  ...          0.0   
...             ...       ...            ...       ...  ...          ...   
1289083    0.036000  0.108000       0.120000      0.00  ...          0.0   
1289493    0.121429  0.100000       0.000000      0.00  ...          0.0   
1289795    0.058824  0.000000       0.000000      0.00  ...          0.0   
1289833    0.045000  0.000000       0.000000      0.05  ...          0.0   
1290429    0.000000  0.236364       0.000000      0.00  ...          0.0   

            Novel  Original     Other  Picture book  Radio  Unknown  \
549      0.000000  0.360000  0.000000           0.0    0.0      0.0   
644      0.000000  0.386667  0.000000           0.0    0.0      0.0   
1519     0.100000  0.137500  0.000000           0.0    0.0      0.0   
2211     0.000000  0.208333  0.000000           0.0    0.0      0.0   
2297     0.000000  0.220000  0.000000           0.0    0.0      0.0   
...           ...       ...       ...           ...    ...      ...   
1289083  0.000000  0.156000  0.000000           0.0    0.0      0.0   
1289493  0.042857  0.114286  0.078571           0.0    0.0      0.0   
1289795  0.000000  0.176471  0.000000           0.0    0.0      0.0   
1289833  0.000000  0.000000  0.000000           0.0    0.0      0.0   
1290429  0.081818  0.000000  0.000000           0.0    0.0      0.0   

         Visual novel  Web manga  Web novel  
549          0.000000        0.1        0.0  
644          0.000000        0.0        0.0  
1519         0.000000        0.0        0.0  
2211         0.050000        0.0        0.0  
2297         0.000000        0.0        0.0  
...               ...        ...        ...  
1289083      0.072000        0.0        0.0  
1289493      0.000000        0.0        0.0  
1289795      0.094118        0.0        0.0  
1289833      0.145000        0.0        0.0  
1290429      0.154545        0.0        0.0  

[3050 rows x 67 columns]

## Now we have User Based Data and Item Based Data so we can predict ratings using User Based Data and we can recommend animes using Item Based Data

#### Lets start with making KNN

I decided to use Jotlib to Parallelize CPU and make faster predictions, and numba for faster numpy calculations because when I tried on kaggle it was so slow, even 1 fold of cross validation lasted 1 hour and 54 minutes. So, it makes the calculation much and much faster.

In [16]:
from tqdm.notebook import tqdm
from joblib import Parallel, delayed
from numba import jit

@jit(nopython=True)
def dot(x1, x2):
    '''
        Paramaters
        _____________
        x1, x2 : np.array, np.array
            Numpy arrays to be multiplied
            
        Description
        ______________
        as Numba's implementation of np.dot doesn't work well on contigous arrays, this function helps dot product.
    '''
    x1 = np.ascontiguousarray(x1)
    x2 = np.ascontiguousarray(x2)
    return np.dot(x1, x2)

@jit(nopython=True)
def cosine_similarity(x1, x2):
    '''
        Paramaters
        _____________
        x1, x2 : np.array, np.array
            Numpy arrays that be looked similarities
            
        Description
        ______________
        Cosine Similarity Function
    '''
    dot_product = dot(x1, x2)
    norm_x1 = np.linalg.norm(x1)
    norm_x2 = np.linalg.norm(x2)
    similarity = dot_product / (norm_x1 * norm_x2)
    return similarity

@jit(nopython=True)
def k_nearest_indices(X, x, k):
    '''
        Paramaters
        _____________
        X : np.array
            This matrix is the searched matrix.

        x : np.array
            This vector is the query vector.
            
        k : int
            how many nearest indices are asked.
        Description
        ______________
        This function helps to find nearest indices of a query
    '''
    similarities = np.array([-cosine_similarity(x, x1) for x1 in X])
    sorted_indices = np.argsort(similarities)
    return sorted_indices[:k]

class KNN():
    '''
        Parameters
        ___
        
        k : int
            number of neighbours will be calculated
        
        metric : str | int
            distance metric that wanted be utilized.
            
        weighted : bool
            decides if KNN is weighted
            
        task : str
            declares the task (classification, regression, recommendation)
    '''
    
    def __init__(self,
                 k: int,
                 weighted : bool,
                 task : str,
                ) -> None:
        
        self.k = k
        self.weighted = weighted
        self.task = task
    def fit(self, X, y):
        '''
            Fits the X and y
            
            Parameters
            ___
            
            X : pd.DataFrame
                input dataframe
            y: pd.DataFrame
                labels
        '''
        self.X = X
        self.y = y
        
    def _predict(self, x):
        '''
            Description
            _______________
            
            Looks k nearest neighbors and predict the class
        '''
        nearest_indices = k_nearest_indices(self.X, x, self.k)
        nearest_similarities = np.array([cosine_similarity(x, self.X[idx]) for idx in nearest_indices])

        if self.weighted:
            labels = {l: np.sum(nearest_similarities[np.where(self.y[nearest_indices] == l)]) for l in np.unique(self.y[nearest_indices])}
        else:
            labels = {l: len(np.where(self.y[nearest_indices] == l)[0]) for l in np.unique(self.y[nearest_indices])}

        return max(labels, key=labels.get)

    def regression(self, x):
        '''
            Description
            ______
            
            Look k nearest neighbors and make a regression.
        '''
        nearest_indices = k_nearest_indices(self.X, x, self.k)
        nearest_labels = self.y[nearest_indices]

        if self.weighted:
            similarities = np.array([cosine_similarity(x, self.X[idx]) for idx in nearest_indices])
        else:
            similarities = np.ones(len(nearest_indices))

        summation = np.sum(similarities)
        prediction = np.sum(nearest_labels * similarities) / summation if summation != 0 else 0
        return prediction
    
    def recommend(self, x):
        '''
        Description
        ________________

        We recommend most similar y values related to input data x.
        '''
        nearest_indices = k_nearest_indices(self.X, x, self.k)
        nearest_labels = self.y[nearest_indices]
        return nearest_labels
    
    def predict(self, X):
        '''
            Description
            ______________
            
            Predict a query matrix.
        '''
        if self.task == 'classification':
            predicted_labels = Parallel(n_jobs=-1)(delayed(self._predict)(x) for x in tqdm(X, desc="Predicting"))
            return np.array(predicted_labels)
        elif self.task == 'regression':
            regressions = Parallel(n_jobs=-1)(delayed(self.regression)(x) for x in tqdm(X, desc="Predicting"))
            return np.array(regressions)
        elif self.task == 'recommendation':
            recommendations = Parallel(n_jobs=-1)(delayed(self.recommend)(x) for x in tqdm(X, desc="Predicting"))
            return recommendations
    '''staticmethod
    @jit(nopython=True)
    def cosine_similarity(self, x1, x2):
        dot_product = np.dot(x1, x2)
        norm_x1 = np.linalg.norm(x1)
        norm_x2 = np.linalg.norm(x2)
        similarity = dot_product / (norm_x1 * norm_x2)
        return similarity
    @staticmethod
    @jit(nopython=True)
    def k_nearest_indices(self, x):
        similarities = np.array([-self.cosine_similarity(x, x1) for x1 in self.X])
        sorted_indices = np.argsort(similarities)
        return sorted_indices[:self.k]

    '''

In [17]:
train_X, train_y = train_merged, train_merged["rating"]
test_X, test_y = test_merged, test_merged["rating"]

In [18]:
from sklearn.preprocessing import StandardScaler

In [19]:
merged_X = pd.concat([train_X, test_X])
scaler = StandardScaler().fit(merged_X)
train_X = scaler.transform(train_X)
test_X = scaler.transform(test_X)

In [20]:
def MAE(y_true, y_pred):
    '''
        Parameters
        ______________
        y_true, y_pred : np.array, np.array
            True and Predicted vectors will be looked at L1 distance.
            
        Description
        _______________
        Give a number that shows error, looking at L1 distance.
    '''
    diff = abs(y_true - y_pred)
    mae = np.mean(diff)
    return mae

def cv(k, model, X, y, shuffle=True, random_state = 42):
    '''
        Parameters
        _______________
        k : int
            how many folds will be done
            
        model
            model will give predictions
            
        X : np.array
            Feature matrix
        
        y : np.array
            Label vector
            
        shuffle : bool
            shuffles X, y if true
            
        random_state : int
            seed to give same random numbers each time.
    '''
    # Shuffle X and y
    if shuffle:
        np.random.seed(random_state)
        perm = np.random.permutation(len(X))
    X, y = X[perm], y[perm]
    errors = []
    # validate
    fold_size = len(X) // k
    for i in range(k):
        start = i * fold_size
        end = (i+1) * fold_size if i != k-1 else len(X)
        
        X_train = np.concatenate((X[:start], X[end:]))
        y_train = np.concatenate((y[:start], y[end:]))
        
        X_val = X[start:end]
        y_val = y[start:end]
        
        model.fit(X_train, y_train)
        y_pred = model.predict(X_val)
        
        mae = MAE(y_val, y_pred)
        
        errors.append(mae)
    return np.mean(errors)

In [43]:
'''errors = []
for k in [3, 5, 7]:
    knn = KNN(k=k, weighted=False, task="regression")
    error = cv(k=5, model=knn, X=train_X, y=train_y)
    errors.append(error)'''
errors = [0.20233095623987034, 0.24104943273905993, 0.2728988191711044]
errors

[0.20233095623987035, 0.24104943273905993, 0.2728988191711044]

![result](pictures/Screenshot%20from%202023-10-30%2011-03-57.png)

In [22]:
'''weighted_errors = []
for k in [3, 5, 7]:
    knn = KNN(k=k, weighted=True, task="regression")
    error = cv(k= 5, model=knn, X=train_X, y=train_y)
    weighted_errors.append(error)'''
weighted_errors = [0.2021547042864606, 0.2406848509179383, 0.27243754988611396]
weighted_errors

[0.2021547042864606, 0.2406848509179383, 0.27243754988611396]

![result2](pictures/Screenshot%20from%202023-10-30%2011-04-07.png)

In [23]:
# Optimal KNN
knn = KNN(k=[3, 5, 7][np.argmin(errors)], weighted=False, task='regression')
knn.fit(train_X, train_y)
predictions = knn.predict(test_X)
quantized_predictions = [round(p) for p in predictions]
MAE(test_y, predictions), MAE(test_y, quantized_predictions)

Predicting:   0%|          | 0/877 [00:00<?, ?it/s]

(0.12086659064994298, 0.10262257696693272)

In [24]:
# Optimal Weighted KNN
wknn = KNN(k=[3, 5, 7][np.argmin(weighted_errors)], weighted=True, task='regression')
wknn.fit(train_X, train_y)
wpredictions = knn.predict(test_X)
quantized_wpredictions = [round(p) for p in wpredictions]
MAE(test_y, wpredictions), MAE(test_y, quantized_wpredictions)

Predicting:   0%|          | 0/877 [00:00<?, ?it/s]

(0.12086659064994298, 0.10262257696693272)

In [25]:
# We can See that MAE of Weighted KNN and KNN is the same. 
# We can conclude probably they are same. So, lets check it.
all(predictions == wpredictions)

True

In [38]:
# As predictions are the same, quantized versions must be same
quantized_predictions == quantized_wpredictions

True

## Evaluation Part

In [26]:
from sklearn import metrics

In [27]:
metrics.accuracy_score(test_y, quantized_wpredictions)

0.9167616875712656

In [39]:
metrics.precision_score(test_y, quantized_predictions, average='micro'), metrics.precision_score(test_y, quantized_predictions, average='macro')

/home/grkmkola/miniconda3/envs/ml/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


(0.9167616875712656, 0.6568678298638589)

In [40]:
metrics.recall_score(test_y, quantized_predictions, average='micro'), metrics.recall_score(test_y, quantized_predictions, average='macro'), 

(0.9167616875712656, 0.6823150997365476)

In [41]:
metrics.f1_score(test_y, quantized_predictions, average='micro'), metrics.f1_score(test_y, quantized_predictions, average='macro')

(0.9167616875712656, 0.6467874945499182)

As we can see on the scores, all accuracy is %91.7 and micro precision, recall and f1 score are 0.917 are the same. As micro averaging gives the same weight each inctance, we can conclude that KNN model is shown high performance on instance level, but on the other hand, macro averaging gives range to (0.64 to 0.68) and it shows imbalance between classes. 

## Recommendation Part

In [31]:
merged_users = pd.Series({**train_user_names, **test_user_names})
merged_users

549                  dotGif
644        zimdudealchemist
1519              wretchedx
2211                   Synn
2297            Yakito-chan
                 ...       
601187              Vitiare
1061301          icewitch92
1123463            Mariotte
1204987       FrozenCthulhu
1257993                 LHT
Length: 3050, dtype: object

In [32]:
def recommendation(model, query, k: int, weighted, data_X, data_y):
    scaler = StandardScaler().fit(data_X)
    data_X = scaler.transform(data_X)
    users = merged_users.loc[query.index]
    query = scaler.transform(query)
    model = model(k=k, weighted=weighted, task='recommendation')
    model.fit(data_X, data_y)
    
    return pd.Series(dict(zip(users, model.predict(query))))

In [33]:
recs = recommendation(KNN, item_based_matrix, 25, True, recommend_X, recommend_y)
recs

Predicting:   0%|          | 0/3050 [00:00<?, ?it/s]

dotGif              1431                    Code Geass: Hangyaku n...
zimdudealchemist    20                    Neon Genesis Evangelion
...
wretchedx           72                             Kidou Senshi Gu...
Synn                100                                  Fullmetal...
Yakito-chan         101                    Full Moon wo Sagashite
...
                                          ...                        
terrybro            24                                        Kouk...
foxx278             227                                        Kyo...
MJVirgo777          1589           Amaenaide yo!! Katsu!!: Damasar...
kuks08              166                                           ...
Anka21              11751                          Ajin Part 2
109...
Length: 3050, dtype: object

In [34]:
def get_recommendation(user):
    if isinstance(user, str):
        return recs[user]
    return recs[merged_users[user]]

In [35]:
# Sample Recommendation by id
get_recommendation(549)

1431                    Code Geass: Hangyaku no Lelouch
2647                 Code Geass: Hangyaku no Lelouch R2
143                                           RahXephon
72                             Kidou Senshi Gundam SEED
1981                                        Dennou Coil
1499                                Uchuu Senkan Yamato
74                                        Turn A Gundam
2469                                     Hidamari no Ki
196       Kidou Senkan Nadesico: The Prince of Darkness
3270                 Evangelion: 3.0 You Can (Not) Redo
2532                Evangelion: 1.0 You Are (Not) Alone
8849     Uchuu Senkan Yamato 2199: Hoshimeguru Hakobune
939             Ouritsu Uchuugun: Honneamise no Tsubasa
100                                 Fullmetal Alchemist
595                                    Legend of Basara
438               Koukaku Kidoutai: Stand Alone Complex
1323                                 Key the Metal Idol
70          Shin Kidou Senki Gundam Wing: Endles

In [36]:
# Sample Recommendation by username
get_recommendation('dotGif')

1431                    Code Geass: Hangyaku no Lelouch
2647                 Code Geass: Hangyaku no Lelouch R2
143                                           RahXephon
72                             Kidou Senshi Gundam SEED
1981                                        Dennou Coil
1499                                Uchuu Senkan Yamato
74                                        Turn A Gundam
2469                                     Hidamari no Ki
196       Kidou Senkan Nadesico: The Prince of Darkness
3270                 Evangelion: 3.0 You Can (Not) Redo
2532                Evangelion: 1.0 You Are (Not) Alone
8849     Uchuu Senkan Yamato 2199: Hoshimeguru Hakobune
939             Ouritsu Uchuugun: Honneamise no Tsubasa
100                                 Fullmetal Alchemist
595                                    Legend of Basara
438               Koukaku Kidoutai: Stand Alone Complex
1323                                 Key the Metal Idol
70          Shin Kidou Senki Gundam Wing: Endles